[feat] Update node output system to use NodeLocatorIds

christian-byrne opened the pull request:
Updates the global node outputs store to use node locator IDs instead of execution IDs using foundation established in https://github.com/Comfy-Org/ComfyUI_frontend/pull/4382.

The `executed` WS message sends messages mapping hierarchical node IDs (based on execution context) to outputs (file references). This PR changes to using locator IDs (immediate parent graph ID : node ID).

During rendering logic, we won't know the execution context of node instances, but we know the node instance's parents graph. So, we can query the store using the active parent graph ID prefixed with the node instance ID.

https://github.com/user-attachments/assets/6225b9dd-ffbb-4b03-99e3-b0dcb10ac65c

┆Issue is synchronized with this [Notion page](https://www.notion.so/PR-4500-feat-Update-node-output-system-to-use-NodeLocatorIds-2396d73d365081af8ceff78e31497af2) by [Unito](https://www.unito.io)

christian-byrne committed:
[feat] Update subgraph output system to use NodeLocatorIds

- Switch imagePreviewStore to use NodeLocatorIds for consistent node identification
- Add setNodeOutputsByExecutionId method that handles hierarchical execution IDs
- Add setNodeOutputsByNodeId method that uses current graph context
- Preserve merge functionality when storing node outputs
- Remove unused imports and simplify implementation

christian-byrne committed:
[refactor] Improve imagePreviewStore implementation

- Remove unnecessary code comments from executed event handler
- Refactor to use shared internal setOutputsByLocatorId function
- Add merge option support to setNodeOutputsByNodeId method
- Share SetOutputOptions interface across all setter methods
- Simplify code by eliminating duplication in output setting logic

christian-byrne committed:
[refactor] Move SetOutputOptions interface to module level

- Move interface definition outside of store function for better code organization
- Follows TypeScript best practices for interface placement

webfiltered reviewed (commented):
Please pause on merging this; I am investigating the use of locator ID here.

christian-byrne commented on src/stores/imagePreviewStore.ts:
```diff
@@ -96,24 +133,67 @@ export const useNodeOutputStore = defineStore('nodeOutput', () => {
+    setOutputsByLocatorId(nodeLocatorId, outputs, options)
+  }
+
+  /**
+   * Set node outputs by node ID.
+   * Uses the current graph context to create the appropriate NodeLocatorId.
+   *
+   * @param nodeId - The node ID
+   * @param outputs - The outputs to store
+   * @param options - Options for setting outputs
+   * @param options.merge - If true, merge with existing outputs (arrays are concatenated)
+   */
+  function setNodeOutputsByNodeId(
+    nodeId: string | number,
+    outputs: ExecutedWsMessage['output'] | ResultItem,
+    options: SetOutputOptions = {}
+  ) {
+    const nodeLocatorId = nodeIdToNodeLocatorId(nodeId)
+    setOutputsByLocatorId(nodeLocatorId, outputs, options)
+  }
```
Yeah, this should be getting called by `setNodeOutputs`.

christian-byrne committed:
[fix] Use setNodeOutputsByNodeId function in setNodeOutputs to address review comment

webfiltered reviewed (commented):
My design concern here is that we are silently discarding the execution ID, which doesn't make a lot of sense to me.

My general design for storing execution results (progress, output, everything) was to always keep the ref to the execution ID, as it is the source of truth.  Not suggesting we don't use locator IDs for the reverse lookup - suggesting both be in the store.

Previews are currently still bust:

<img width="888" height="586" alt="image" src="https://github.com/user-attachments/assets/c36a8838-204b-49ce-802e-3b39138241a5" />

christian-byrne commented:
> My general design for storing execution results (progress, output, everything) was to always keep the ref to the execution ID, as it is the source of truth. Not suggesting we don't use locator IDs for the reverse lookup - suggesting both be in the store.

We can add executionId->nodeOutput to the store but since it won't be used yet it might belong in followup PR. Currently, the output store is also used for things like LoadImage which don't have access to the execution Id.

christian-byrne committed:
[feat] Add graph traversal utilities and update subgraph support

- Add graphTraversalUtil with functions to find nodes anywhere in graph hierarchy
- Update uploadAudio to use NodeLocatorIds and graph traversal utilities
- Fix event listeners in executionStore (add missing cleanup)
- Add preview management functions to imagePreviewStore
- Update queueStore to use nodeOutputsStore for outputs
- Add comprehensive unit tests for graph traversal utilities

This completes the implementation of review feedback for proper subgraph node handling.

christian-byrne committed:
[cleanup] Remove unnecessary comments from graphTraversalUtil

- Remove section header comments
- Remove 'Pure function with no side effects' from JSDoc comments
- Keep only essential documentation
